[SPARK-9023] [SQL] Efficiency improvements for UnsafeRows in Exchange

JoshRosen opened the pull request:
This pull request aims to improve the performance of SQL's Exchange operator when shuffling UnsafeRows.  It also makes several general efficiency improvements to Exchange.

Key changes:
- When performing hash partitioning, the old Exchange projected the partitioning columns into a new row then passed a `(partitioningColumRow: InternalRow, row: InternalRow)` pair into the shuffle. This is very inefficient because it ends up redundantly serializing the partitioning columns only to immediately discard them after the shuffle.  After this patch's changes, Exchange now shuffles `(partitionId: Int, row: InternalRow)` pairs.  This still isn't optimal, since we're still shuffling extra data that we don't need, but it's significantly more efficient than the old implementation; in the future, we may be able to further optimize this once we implement a new shuffle write interface that accepts non-key-value-pair inputs.
- Exchange's `compute()` method has been significantly simplified; the new code has less duplication and thus is easier to understand.
- When the Exchange's input operator produces UnsafeRows, Exchange will use a specialized `UnsafeRowSerializer` to serialize these rows.  This serializer is significantly more efficient since it simply copies the UnsafeRow's underlying bytes.  Note that this approach does not work for UnsafeRows that use the ObjectPool mechanism; I did not add support for this because we are planning to remove ObjectPool in the next few weeks.

JoshRosen committed:
Iniitial cut at removing shuffle on KV pairs

JoshRosen committed:
Big code simplification in Exchange

JoshRosen committed:
Import ordering

JoshRosen committed:
Add UnsafeRowSerializer

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into unsafe-shuffle

JoshRosen committed:
Add logic for choosing when to use UnsafeRowSerializer

JoshRosen committed:
Fix for copying logic

JoshRosen committed:
Exchange outputs UnsafeRows when its child outputs them

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37562 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37562/consoleFull) for   PR 7456 at commit [`8dd3ff2`](https://github.com/apache/spark/commit/8dd3ff21016bf9b5217626c5b9723f43de055def).

SparkQA commented:
[Test build #37562 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37562/console) for   PR 7456 at commit [`8dd3ff2`](https://github.com/apache/spark/commit/8dd3ff21016bf9b5217626c5b9723f43de055def).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ShuffledRowRDD(`
  * `class UnsafeRowSerializer(numFields: Int) extends Serializer `

AmplabJenkins commented:
Merged build finished. Test FAILed.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/UnsafeRowSerializer.scala:
```diff
@@ -0,0 +1,130 @@
+import org.apache.spark.unsafe.PlatformDependent
+
+
+class UnsafeRowSerializer(numFields: Int) extends Serializer {
```
this is a file that deserves a little bit more comment to explain what this is for.

JoshRosen replied:
Yep.  I'm going to add comments now.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/UnsafeRowSerializer.scala:
```diff
@@ -0,0 +1,130 @@
+      val row = value.asInstanceOf[UnsafeRow]
+      assert(row.getPool == null, "UnsafeRowSerializer does not support ObjectPool")
+      dOut.writeInt(row.getSizeInBytes)
+      println("Size in bytes is " + row.getSizeInBytes)
```
make sure you remove this when it's merged (i guess our style checker will fail anyway)

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/Exchange.scala:
```diff
@@ -112,109 +114,71 @@ case class Exchange(newPartitioning: Partitioning, child: SparkPlan) extends Una
-            val mutablePair = new MutablePair[InternalRow, Null]()
-            iter.map(row => mutablePair.update(row, null))
-          }
+        // Internally, RangePartitioner runs a job on the RDD that samples keys to compute
```
ouch total order sort is super inefficient - we should definitely try to make it a lot more efficient in 1.6.

rxin commented:
The current code looks pretty good to me.

JoshRosen commented:
I'm going to rebase this on top of #7482 to make things easier to test; will rebase again once that patch is merged.

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into unsafe-exchange

JoshRosen committed:
Remove println() and add comments

JoshRosen committed:
Add simple test of UnsafeRow shuffling in Exchange.

JoshRosen committed:
Add missing newline

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

JoshRosen committed:
Some additional comments + small cleanup to remove an unused parameter

JoshRosen commented:
Alright, I've updated this and it should be ready for another look.  Added a very trivial test to trigger the new shuffle path and caught a bug related to UnsafeRowSerializer not being Serializable.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37736 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37736/consoleFull) for   PR 7456 at commit [`0082515`](https://github.com/apache/spark/commit/00825153fab88c183b16beeeb83e181846f37518).

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37736 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37736/console) for   PR 7456 at commit [`0082515`](https://github.com/apache/spark/commit/00825153fab88c183b16beeeb83e181846f37518).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Concat(children: Seq[Expression]) extends Expression with ImplicitCastInputTypes `
  * `class ShuffledRowRDD(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #37738 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37738/consoleFull) for   PR 7456 at commit [`0082515`](https://github.com/apache/spark/commit/00825153fab88c183b16beeeb83e181846f37518).

SparkQA commented:
[Test build #37738 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37738/console) for   PR 7456 at commit [`0082515`](https://github.com/apache/spark/commit/00825153fab88c183b16beeeb83e181846f37518).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Concat(children: Seq[Expression]) extends Expression with ImplicitCastInputTypes `
  * `class ShuffledRowRDD(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

rxin commented:
Can you also update the pull request description?

JoshRosen commented:
Done; I removed the "work-in-progress" part.

JoshRosen commented:
Huh, looks like a legitimate test failure in SparkSqlSerializer2SortMergeShuffleSuite:

```
org.apache.spark.rdd.MapPartitionsRDD cannot be cast to org.apache.spark.rdd.ShuffledRDD
```

JoshRosen committed:
Fix cast in SparkSqlSerializer2Suite

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37744 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37744/consoleFull) for   PR 7456 at commit [`7e75259`](https://github.com/apache/spark/commit/7e7525995d383af03a8b4179e8f3139cd97e2e12).

SparkQA commented:
[Test build #37744 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37744/console) for   PR 7456 at commit [`7e75259`](https://github.com/apache/spark/commit/7e7525995d383af03a8b4179e8f3139cd97e2e12).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ShuffledRowRDD(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/UnsafeRowSerializer.scala:
```diff
@@ -0,0 +1,142 @@
+      this
+    }
+    override def writeKey[T: ClassTag](key: T): SerializationStream = {
+      assert(key.isInstanceOf[Int])
```
you need to add some comment explaining why we are not doing anything when writing keys.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/UnsafeRowSerializer.scala:
```diff
@@ -0,0 +1,142 @@
+      var dataRemaining: Int = row.getSizeInBytes
+      val baseObject = row.getBaseObject
+      var rowReadPosition: Long = row.getBaseOffset
+      while (dataRemaining > 0) {
```
probably doesn't matter in the MVP, but if we know the UnsafeRow is backed by a byte array, we don't need to do this copying, do we?

JoshRosen replied:
A nice way to address this would be to add a `writeTo` method to `UnsafeRow` itself.  That method could contain a special case to handle the case where the row is backed by an on-heap byte array.

rxin replied:
+1

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/UnsafeRowSerializer.scala:
```diff
@@ -0,0 +1,142 @@
+  override def deserializeStream(in: InputStream): DeserializationStream = {
+    new DeserializationStream {
+      private[this] val dIn: DataInputStream = new DataInputStream(in)
+      private[this] var rowBuffer: Array[Byte] = new Array[Byte](1024)
```
why is the buffer here only 1024?

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/UnsafeRowSerializer.scala:
```diff
@@ -0,0 +1,142 @@
+
+private class UnsafeRowSerializerInstance(numFields: Int) extends SerializerInstance {
+
+  private[this] val EOF: Int = -1
```
it'd be great to briefly explain the encoding format, i.e. length followed by data, and if length < 0, eof.

rxin commented:
Looks pretty good. I'm going to merge it. Please submit a followup pr to address some of the comments on documentation and choice of buffer size.
